effekt/shared/src/main/scala/effekt/RecursiveDescent.scala

```diff
@@ -1100,78 +1096,251 @@ class RecursiveDescent(positions: Positions, tokens: Seq[Token], source: Source)
+            case tpe :: Nil => tpe
+            case tpes => TypeTuple(tpes)
+          }
+        case _ => refType()
+      }
+
+  // Parse function types (middle precedence)
+  private def functionType(): GeneralType = {
+    // temporary:
+    def parenTypes(): List[GeneralType] =
+      nonterminal:
+        many(boxedType, `(`, `,`, `)`)
+
+    def maybeValueTypes(): List[GeneralType] =
+      nonterminal:
+        if peek(`(`) then parenTypes() else Nil
+
+
+    def maybeBlockTypeParams(): List[(Option[IdDef], GeneralType)] =
+      nonterminal:
+        if peek(`{`) then blockTypeParams() else Nil
+
+    def blockTypeParams(): List[(Option[IdDef], GeneralType)] =
+      nonterminal:
+        someWhile(blockTypeParam(), `{`)
+
+    def blockTypeParam(): (Option[IdDef], GeneralType) =
+      nonterminal:
+        braces {
+          (backtrack { idDef() <~ `:` }, boxedType())
         }
-        case _ => ValueTypeRef(idRef(), maybeTypeArgs())
+
+    // Complex function type: [T]*(Int, String)*{Exc} => Int / {Effect}
+    def functionTypeComplex = backtrack {
+      maybeTypeParams() ~ maybeValueTypes() ~ (maybeBlockTypeParams() <~ `=>`) ~ atomicType() ~ maybeEffects() match {
+        case tparams ~ vparams ~ bparams ~ t ~ effs => TypeFun(tparams, vparams, bparams, t, effs)
       }
+    }
 
+    // Simple function type: Int => Int
+    def functionTypeSimple = backtrack {
+      TypeRef(idRef(), maybeGeneralTypeArgs()) <~ `=>`
+    } map { tpe =>
+      TypeFun(Nil, List(tpe), Nil, atomicType(), maybeEffects())
+    }
 
-  /**
-   * Uses backtracking!
-   *
-   * TODO improve errors
-   *   i.e. fail("Expected either a function type (e.g., (A) => B / {E} or => B) or an interface type (e.g., State[T]).")
-   */
-  def blockType(): BlockType = blockType2(true)
-  private def blockType2(boxedAllowed: Boolean): BlockType =
+    // Try to parse each function type variant, fall back to basic type if none match
     nonterminal:
+      functionTypeSimple orElse functionTypeComplex getOrElse atomicType()
+  }
 
-      def simpleFunType = backtrack {
-        ValueTypeRef(idRef(), maybeTypeArgs()) <~ `=>`
-      } map { tpe =>
-        FunctionType(Nil, List(tpe), Nil, valueType2(boxedAllowed), maybeEffects())
+  // Parse boxed types (lowest precedence)
+  private def boxedType(): GeneralType = {
+    nonterminal:
+      // Parse the function type first
+      val tpe = functionType()
+
+      val boxed = when(`at`) {
+        TypeBox(tpe, captureSet())
+      } {
+        tpe
       }
 
-      def funType = backtrack {
-        maybeTypeParams() ~ maybeValueTypes() ~ (maybeBlockTypeParams() <~ `=>`) ~ valueType2(boxedAllowed) ~ maybeEffects() match {
-          case tparams ~ vparams ~ bparams ~ t ~ effs => FunctionType(tparams, vparams, bparams, t, effs)
-        }
+      if (peek(`/`)) {
+        TypeEff(boxed, maybeEffects())
+      } else {
+        boxed
       }
-      def parenthesized = backtrack { parens { blockType() } }
+  }
```

Comment:
Technically both of the `at {...}` parsing and `/ ...` parsing should be while loops / recursive functions to catch the most errors possible like `foo at {...} at {...} at {...}`, but I'm somewhat tired of parsing shenanigans, so we can keep it for a future issue. :)